tower-middleware/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+//! This additional functionality may include, but is not limited to:
+//!
+//! * Rejecting the request.
+//! * Taking an action based on the request.
+//! * Mutating the request before passing it along to the application.
+//! * Mutating the response returned by the application.
+//!
+//! A middleware implements the [`Middleware`] trait.
+
+#![deny(missing_docs)]
+#![doc(html_root_url = "https://docs.rs/tower-middleware/0.1.0")]
+
+extern crate futures;
+extern crate tower_service;
+
+#[cfg(feature = "ext")]
+mod ext;
+
+#[cfg(feature = "ext")]
+pub use ext::{Chain, MiddlewareExt};
+
+use tower_service::Service;
+
+/// Decorates a `Service`, transforming either the request or the response.
+///
+/// Often, many of the pieces needed for writing network applications can be
+/// reused across multiple services. The `Middleware` trait can be used to write
+/// reusable components that can be applied to very different kinds of services;
+/// for example, it can be applied to services operating on different protocols,
+/// and to both the client and server side of a network transaction.
+///
+/// # Log
+///
+/// Take request logging as an example:
+///
+/// ```rust
+/// # extern crate futures;
+/// # extern crate tower_service;
+/// # extern crate tower_web;
+/// # #[macro_use] extern crate log;
+/// use tower_service::Service;
+/// use tower_web::middleware::Middleware;
+/// use futures::{Future, Poll};
+///
+/// use std::fmt;
+///
+/// pub struct LogMiddleware {
+///     target: &'static str,
+/// }
+///
+/// impl<S> Middleware<S> for LogMiddleware
+/// where
+///     S: Service,
+///     S::Request: fmt::Debug,
+/// {
+///     type Request = S::Request;
+///     type Response = S::Response;
+///     type Error = S::Error;
+///     type Service = LogService<S>;
+///
+///     fn wrap(&self, service: S) -> LogService<S> {
+///         LogService {
+///             target: self.target,
+///             service
+///         }
+///     }
+/// }
+///
+/// // This service implements the Log behavior
+/// pub struct LogService<S> {
+///     target: &'static str,
+///     service: S,
+/// }
+///
+/// impl<S> Service for LogService<S>
+/// where
+///     S: Service,
+///     S::Request: fmt::Debug,
+/// {
+///     type Request = S::Request;
+///     type Response = S::Response;
+///     type Error = S::Error;
+///     type Future = S::Future;
+///
+///     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+///         self.service.poll_ready()
+///     }
+///
+///     fn call(&mut self, request: Self::Request) -> Self::Future {
+///         info!(target: self.target, "request = {:?}", request);
+///         self.service.call(request)
+///     }
+/// }
+/// ```
+///
+/// The above log implementation is decoupled from the underlying protocol and
+/// is also decoupled from client or server concerns. In other words, the same
+/// log middleware could be used in either a client or a server.
+pub trait Middleware<S> {
```

Comment:
Should `S` be constrained? i.e.`Middleware<R, S: Service<R>>` ?  My hunch is that lacking this constraint will cause some discomfort for implementors. But I'd have to think a bit more to articulate that

Reply:
I don't think we can constrain it due to coherence. Or, if we can, it is just a huge PITA to do so w/o *much* win.

Reply:
I'm not sure but I had to constrain it here but afaik I would need to continue to use constraints on all the impls even if it is on the core trait or not.

https://github.com/LucioFranco/tower/blob/service-builder/tower-in-flight-limit/src/lib.rs#L148

Reply:
Yeah, it's probably  not worth pursuing...